pyccel/stdlib/ndarrays/ndarrays.c

```diff
@@ -373,3 +429,65 @@ int64_t     *numpy_to_ndarray_shape(int64_t *np_shape, int nd)
     return nd_shape;
 
 }
+
+bool    is_same_shape(t_ndarray dest, t_ndarray src)
+{
+    int i = 0;
+
+    if (dest.nd != src.nd)
+        return false;
+    while (i < dest.nd)
+    {
+        if (dest.shape[i] != src.shape[i])
+            return false;
+        ++i;
+    }
+    return true;
+}
+
+int get_shape_product(int64_t *shape, int nd, int max_nd)
+{
+    int product = 1;
+
+    for (int i = nd; i < max_nd; ++i)
+        product *= shape[i];
+    return (product);
+}
+
+int index_position(t_ndarray arr, int index, int nd)
+{
+    if (arr.order == order_c)
+        return index;
+    if (nd == 0)
+        return (0);
+    if (nd == arr.nd)
+        return (index % arr.shape[nd - 1]) * arr.strides[nd - 1] + index_position(arr, index, nd - 1);
+    int true_index = (index / (get_shape_product(arr.shape, nd, arr.nd)));
+    if (true_index >= arr.shape[nd - 1])
+        true_index = true_index % arr.shape[nd - 1];
+    return (true_index * arr.strides[nd - 1] + index_position(arr, index, nd - 1));
+}
+
+void array_copy_data(t_ndarray dest, t_ndarray src, int64_t offset)
+{
+    unsigned char *d = (unsigned char*)dest.raw_data;
+    unsigned char *s = (unsigned char*)src.raw_data;
+
+    if (dest.order == src.order)
+    {
+        if ((dest.order == order_c) || (dest.order == order_f && is_same_shape(dest, src)))
+        {
+            memcpy(d + offset * dest.type_size, s, src.buffer_size);
+            return ;
+        }

```

Comment:
```
else
{
    for (int i=0; i<src.length; ++i)
    {
        int flat_src_idx = 0;
        int src_idx = i;
        for (int n=src.nd-1; n>-1; --i)
        {
             int n_th_idx = src_idx%src.shape[n];
             flat_src_idx += src.strides[n]*n_th_idx;
             src_idx -= n_th_idx;
             src_idx /= src.shape[n];
        }
        dest[offset + idx] = src[i];
    }
}
```
This needs testing and there may be possible improvements but I think this is what's missing. Would it make sense to save `strides` and `f_strides` in `t_ndarray`?

Reply:
Yes it might be useful to have both of them, but can't we just do the calculations if needed? using the transpose function maybe, it does create a view, so maybe that's too much.
as for line 480, that is just an optimization in case we don't need to do any calculations and just copy the whole array in one go, using the highly efficient `memcpy`, the for loop calculation is what we do in `index_position`

Reply:
If they are not too costly then yes. I was just concerned that the index calculation might be more costly to calculate than an `f_strides` array?

I agree that the `memcpy` is better if it can be used. That's why I put an `else`. I realise now that I missed the `return` after the `memcpy`